timemachine/md/smc.py

```diff
@@ -0,0 +1,159 @@
+BatchLogProb = Callable[[Samples, Lambda], LogWeights]
+
+Resampler = Callable[[LogWeights], Tuple[IndexArray, LogWeights]]
+ResultDict = Dict[str, Any]
+
+
+def sequential_monte_carlo(
+    samples: Samples,
+    lambdas: Array,
+    propagate: BatchPropagator,
+    log_prob: BatchLogProb,
+    resample: Resampler,
+) -> ResultDict:
+    """barebones implementation of Sequential Monte Carlo (SMC)
+
+    Parameters
+    ----------
+    samples: [N,] list
+    lambdas: [K,] array
+    propagate: function
+        [move(x, lam) for x in xs]
+        for example, move(x, lam) might mean "run 100 steps of all-atom MD targeting exp(-u(., lam)), initialized at x"
+    log_prob: function
+        [exp(-u(x, lam)) for x in xs]
+    resample: function
+        (optionally) perform resampling given an array of log weights
+
+    Returns
+    -------
+    trajs_dict
+        "sample_traj"
+            [K-1, N] list of snapshots
+        "incremental_log_weights_traj"
+            [K-1, N] array of incremental log weights
+        "ancestry_traj"
+            [K-1, N] array of ancestor idxs
+        "log_weights_traj"
+            [K, N] array of accumulated log weights
+
+    References
+    ----------
+    * Arnaud Doucet's annotated bibliography of SMC
+        https://www.stats.ox.ac.uk/~doucet/smc_resources.html
+    * [Dai, Heng, Jacob, Whiteley, 2020] An invitation to sequential Monte Carlo samplers
+        https://arxiv.org/abs/2007.11936
+    """
+    n = len(samples)
+    log_weights = np.zeros(n)
+
+    # store
+    sample_traj = [samples]
+    ancestry_traj = [np.arange(n)]
+    log_weights_traj = [np.array(log_weights)]
+    incremental_log_weights_traj = []  # note: redundant but convenient
+
+    trange = tqdm(lambdas[:-2])
+
+    def accumulate_results(samples, indices, log_weights, incremental_log_weights):
+        sample_traj.append(samples)
+        ancestry_traj.append(indices)
+        log_weights_traj.append(np.array(log_weights))
+        incremental_log_weights_traj.append(np.array(incremental_log_weights))
+        running_estimate = -logsumexp(log_weights - np.log(len(log_weights)))
+        trange.set_postfix(EXP=running_estimate)
+
+    # main loop
+    for (lam_initial, lam_target) in zip(trange, lambdas[1:-1]):
+        # update log weights
+        incremental_log_weights = log_prob(sample_traj[-1], lam_target) - log_prob(sample_traj[-1], lam_initial)
+        log_weights += incremental_log_weights
+
+        # resample
+        indices, log_weights = resample(log_weights)
+        resampled = [sample_traj[-1][i] for i in indices]
+
+        # propagate
+        samples = propagate(resampled, lam_target)
+
+        # log
+        accumulate_results(samples, indices, log_weights, incremental_log_weights)
+
+    # final result: a collection of samples, with associated log weights
+    incremental_log_weights = log_prob(samples, lambdas[-1]) - log_prob(samples, lambdas[-2])
```

Comment:
samples here are collected from lambda[-2]. Should propagate be run at lambdas[-1] to add to sample_traj? (Depends on answer to above question)

Reply:
They're proposed according to `p(x | lambda[-2])`, but weighted for the difference between `p(x | lambda[-1])` and `p(x | lambda[-2])`